tests/integration/psync2.tcl

```diff
@@ -154,21 +154,59 @@ start_server {} {
             fail "Replica did not inherit the new replid."
         }
 
-        # 2) Attach all the slaves to a random instance
+        # Build a lookup with the direct connection master of each replica.
+        # First loop that uses random to decide who replicates from who.
+        array set slave_to_master {}
         while {[llength $used] != 5} {
             while 1 {
                 set slave_id [randomInt 5]
                 if {[lsearch -exact $used $slave_id] == -1} break
             }
             set rand [randomInt [llength $used]]
             set mid [lindex $used $rand]
-            set master_host $R_host($mid)
-            set master_port $R_port($mid)
+            set slave_to_master($slave_id) $mid
+            lappend used $slave_id
+        }
 
-            test "PSYNC2: Set #$slave_id to replicate from #$mid" {
-                $R($slave_id) slaveof $master_host $master_port
+        # 2) Attach all the slaves to a random instance
+        # Second loop that dose the actual SLAVEOF command and make sure execute it in the right order.
+        while {[array size slave_to_master] > 0} {
+            foreach slave_id [array names slave_to_master] {
+                set mid $slave_to_master($slave_id)
+
+                # We only attach the replica to a random instance that already in the old/new chain.
+                if {$root_master($mid) == $root_master($master_id)} {
+                    # Find a replica that can be attached to the new chain already attached to the new master.
+                    # My new master is in the new chain.
+                } elseif {$root_master($mid) == $root_master($slave_id)} {
+                    # My new master and i are in the old chain.
+                } else {
+                    # In cycle 1, we do not care the order.
+                    if {$cycle != 1} {
+                        test "PSYNC2: $slave_id ($root_master($slave_id)) slaveof $mid ($root_master($mid)) continue" {}
```

Comment:
this is here just to print?
looks a bit odd. maybe we can just convert that to a comment. or if you want the debug message, you can add a `puts` inside an `if {$::verbose}`
```suggestion
                        # skipping this replica for now to avoid attaching in a bad order
                        # this is done to avoid an unexpected full sync, when we take a
                        # replica that already reconnected to the new chain and got a new replid
                        # and is then set to connect to a master that's still not aware of that new replid
```

Reply:
Yeah, it's a little weird. My previous thought was that we could see the skip in the log and verify that it did exist. Somehow ending like this.. I will take this comment